modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/AndesRecoveryTask.java

```diff
@@ -23,129 +23,197 @@
-            List<AndesBinding> fromDB = andesContextStore.getBindingsStoredForExchange(exchange.exchangeName);
-            List<AndesBinding> fromMap = amqpConstructStore.getBindingsForExchange(exchange.exchangeName);
-            List<AndesBinding> fromDBDuplicated = new ArrayList<AndesBinding>(fromDB);
-            fromDB.removeAll(fromMap);
-            for (AndesBinding binding : fromDB) {
-                for (BindingListener listener : bindingListeners) {
-                    log.warn("Recovering node. Adding binding " + binding.toString());
-                    listener.handleClusterBindingsChanged(binding, BindingListener.BindingEvent.ADDED);
-                }
-            }
-
-            fromMap.removeAll(fromDBDuplicated);
-            for (AndesBinding binding : fromMap) {
-                for (BindingListener listener : bindingListeners) {
-                    log.warn("Recovering node. removing binding " + binding.toString());
-                    listener.handleClusterBindingsChanged(binding, BindingListener.BindingEvent.DELETED);
-                }
-            }
-        }
-    }
-
-    private void reloadSubscriptions() throws AndesException {
-        ClusterResourceHolder.getInstance().getSubscriptionManager().reloadSubscriptionsFromStorage();
-    }
+public class AndesRecoveryTask implements Runnable, StoreHealthListener {
+
+	private List<QueueListener> queueListeners = new ArrayList<>();
+	private List<ExchangeListener> exchangeListeners = new ArrayList<>();
+	private List<BindingListener> bindingListeners = new ArrayList<>();
+
+	AndesContextStore andesContextStore;
+	AMQPConstructStore amqpConstructStore;
+
+	// set storeOperational to true since it can be assumed that the store is operational at startup
+	// if it is non-operational, the value will be updated immediately
+	AtomicBoolean isContextStoreOperational = new AtomicBoolean(true);
+
+	private static Log log = LogFactory.getLog(AndesRecoveryTask.class);
+
+	public AndesRecoveryTask() {
+
+		// Register AndesRecoveryTask class as a StoreHealthListener
+		FailureObservingStoreManager.registerStoreHealthListener(this);
+
+		queueListeners.add(new ClusterCoordinationHandler(HazelcastAgent.getInstance()));
+		exchangeListeners.add(new ClusterCoordinationHandler(HazelcastAgent.getInstance()));
+		bindingListeners.add(new ClusterCoordinationHandler(HazelcastAgent.getInstance()));
+
+		andesContextStore = AndesContext.getInstance().getAndesContextStore();
+		amqpConstructStore = AndesContext.getInstance().getAMQPConstructStore();
+	}
+
+	@Override
+	public void run() {
+		if(isContextStoreOperational.get()) {
+			try {
+				reloadExchangesFromDB();
+				reloadQueuesFromDB();
+				reloadBindingsFromDB();
+				reloadSubscriptions();
+			} catch (AndesException e) {
+				log.error("Error in running andes recovery task", e);
+			}
+		} else {
+			log.info("AndesRecoveryTask was paused due to non-operational context store.");
+		}
+	}
```

Comment:
This will run periodically to bring the node to the current state of cluster in case some hazelcast notifications are missed.